console/src/app/components/features/features.component.ts

```diff
@@ -78,140 +98,59 @@ export class FeaturesComponent {
-              : !!this.featureData.oidcTriggerIntrospectionProjections?.enabled
-                ? ToggleState.ENABLED
-                : ToggleState.DISABLED,
-        },
-        oidcLegacyIntrospection: {
-          source: this.featureData.oidcLegacyIntrospection?.source || Source.SOURCE_SYSTEM,
-          state:
-            this.featureData.oidcLegacyIntrospection?.source === Source.SOURCE_SYSTEM ||
-            this.featureData.oidcLegacyIntrospection?.source === Source.SOURCE_UNSPECIFIED
-              ? ToggleState.INHERITED
-              : !!this.featureData.oidcLegacyIntrospection?.enabled
-                ? ToggleState.ENABLED
-                : ToggleState.DISABLED,
-        },
-        userSchema: {
-          source: this.featureData.userSchema?.source || Source.SOURCE_SYSTEM,
-          state:
-            this.featureData.userSchema?.source === Source.SOURCE_SYSTEM ||
-            this.featureData.userSchema?.source === Source.SOURCE_UNSPECIFIED
-              ? ToggleState.INHERITED
-              : !!this.featureData.userSchema?.enabled
-                ? ToggleState.ENABLED
-                : ToggleState.DISABLED,
-        },
-        oidcTokenExchange: {
-          source: this.featureData.oidcTokenExchange?.source || Source.SOURCE_SYSTEM,
-          state:
-            this.featureData.oidcTokenExchange?.source === Source.SOURCE_SYSTEM ||
-            this.featureData.oidcTokenExchange?.source === Source.SOURCE_UNSPECIFIED
-              ? ToggleState.INHERITED
-              : !!this.featureData.oidcTokenExchange?.enabled
-                ? ToggleState.ENABLED
-                : ToggleState.DISABLED,
-        },
-        actions: {
-          source: Source.SOURCE_SYSTEM,
-          state:
-            this.featureData.actions?.source === Source.SOURCE_SYSTEM ||
-            this.featureData.actions?.source === Source.SOURCE_UNSPECIFIED
-              ? ToggleState.INHERITED
-              : !!this.featureData.actions?.enabled
-                ? ToggleState.ENABLED
-                : ToggleState.DISABLED,
-        },
-        oidcSingleV1SessionTermination: {
-          source: this.featureData.oidcSingleV1SessionTermination?.source || Source.SOURCE_SYSTEM,
-          state:
-            this.featureData.oidcSingleV1SessionTermination?.source === Source.SOURCE_SYSTEM ||
-            this.featureData.oidcSingleV1SessionTermination?.source === Source.SOURCE_UNSPECIFIED
-              ? ToggleState.INHERITED
-              : !!this.featureData.oidcSingleV1SessionTermination?.enabled
-                ? ToggleState.ENABLED
-                : ToggleState.DISABLED,
-        },
+  private createToggleStates(featureData: GetInstanceFeaturesResponse): ToggleStates {
+    const toggleStates: Partial<ToggleStates> = {};
+
+    FEATURE_KEYS.forEach((key) => {
+      // TODO: Fix this type cast as not all keys are present as FeatureFlag
+      const feature = featureData[key] as unknown as FeatureFlag.AsObject;
```

Comment:
i don't think the type is as FeatureFlag.AsObject because this is from the old grpc client.

Reply:
I think if you do add **as const** above in the declaration of FEATURE_KEYS. Typescript should be able to figure out that your featureData keys always return an FeatureFlag so no casting would be needed.

Reply:
nice catch.. I have also change Source 👍